pkg/provisioners/gcppubsub/dispatcher/dispatcher/reconcile.go

```diff
@@ -0,0 +1,331 @@
+	channelKey := key(c)
+	activeSubscribers := r.subscriptions[channelKey]
+	if len(subscribers.Subscribers) == len(activeSubscribers) {
+		return nil
+	}
+
+	// subsToDelete is logically a set, not a map (values have no meaning).
+	subsToDelete := map[subscriptionName]bool{}
+	for sub := range activeSubscribers {
+		subsToDelete[sub] = true
+	}
+	for _, sub := range subscribers.Subscribers {
+		delete(subsToDelete, subscriptionKey(&sub))
+	}
+	for subToDelete := range subsToDelete {
+		r.subscriptions[channelKey][subToDelete]()
+	}
+	return nil
+}
+
+// createSubscriptionUnderLock starts a background Goroutine for a single subscriber polling its
+// GCP PubSub Subscription.
+// Note that it can only be called if reconciler.subscriptionsLock is held.
+func (r *reconciler) createSubscriptionUnderLock(ctx context.Context, c *eventingv1alpha1.Channel, sub *v1alpha1.ChannelSubscriberSpec) error {
+	ctxWithCancel, cancelFunc := context.WithCancel(ctx)
+
+	channelKey := key(c)
+	if r.subscriptions[channelKey] == nil {
+		r.subscriptions[channelKey] = map[subscriptionName]context.CancelFunc{}
+	}
+	subKey := subscriptionKey(sub)
+	if r.subscriptions[channelKey][subKey] != nil {
+		// There is already a Goroutine watching this subscription.
+		return nil
+	}
+	r.subscriptions[channelKey][subKey] = cancelFunc
+
+	gcpProject := r.defaultGcpProject
+	creds, err := pubsubutil.GetCredentials(ctx, r.client, r.defaultSecret, r.defaultSecretKey)
+	if err != nil {
+		return err
+	}
+	psc, err := r.pubSubClientCreator(ctxWithCancel, creds, gcpProject)
+	if err != nil {
+		return err
+	}
+
+	// receiveMessageBlocking blocks, so run it in a goroutine.
+	go r.receiveMessagesBlocking(ctxWithCancel, c, sub.DeepCopy(), gcpProject, psc)
+
+	return nil
+}
+
+// receiveMessagesBlocking receives messages from GCP PubSub, while blocking forever. If the receive
+// fails for any reason, then it will instruct the reconciler to process this Channel again via
+// reconciler.reconcileChan.
+func (r *reconciler) receiveMessagesBlocking(ctxWithCancel context.Context, c *eventingv1alpha1.Channel, sub *v1alpha1.ChannelSubscriberSpec, gcpProject string, psc pubsubutil.PubSubClient) {
+	subscription := psc.SubscriptionInProject(pubsubutil.GenerateSubName(sub), gcpProject)
+	defaults := provisioners.DispatchDefaults{
+		Namespace: c.Namespace,
+	}
+	channelKey := key(c)
+	subKey := subscriptionKey(sub)
+
+	logging.FromContext(ctxWithCancel).Info("subscription.Receive start")
+	receiveErr := subscription.Receive(
+		ctxWithCancel,
+		receiveFunc(logging.FromContext(ctxWithCancel), sub, defaults, r.dispatcher))
+	// We want to minimize holding the lock. r.reconcileChan may block, so definitely do not do
+	// it under lock. But, to prevent a race condition, we must delete from r.subscriptions
+	// before using r.reconcileChan.
+	func() {
+		r.subscriptionsLock.Lock()
+		defer r.subscriptionsLock.Unlock()
+		delete(r.subscriptions[channelKey], subKey)
```

Comment:
What is the race condition here? Seems like it would be preferable to wait and let the next syncSubscriptions clean up the subscription - no risk of deadlock and the cleanup code is only in one place. But I'm guessing you tried that first :smile:

Reply:
If we didn't `delete(r.subscriptions[channelKey], subKey)`, then on the next `syncSubscriptions`, we would see that `r.subscriptions[channelKey][subKey]` is already present and thus not start a new receiving Goroutine.

This code is the only code that will delete anything in `r.subscriptions[channelKey]`. The pattern I am going for is that the receiving Goroutine cleans itself up after it is finished.

But, this comment made me think about exactly where things get deleted and realized that it is possible for `r.subscriptions[channelKey]` to already have been deleted. I believe everything would work correctly with `delete(r.subscriptions[channelKey], subKey)` (as if the first parameter is `nil` it is a no-op), but added a comment and made it very explicit.